### DISCUSSION TITLE
How to use API generation `access_token`

### DISCUSSION THREAD

**Discussion by bmphjhu906** [zitadel/zitadel#5892] [Q&A]
Hi, guys
I would like to request `/oauth/v2/token`to obtain a `access_token` using Postman. How can I generate `code` and `code_verifier`.

**Comment by mffap** [comment]
You need to login and in the redirect url, you will get the code (assuming you use PKCE). Checkout the [OIDC playground](https://zitadel.com/docs/apis/openidoauth/authrequest) on how to set up the auth request, including code challenge. Haven't tried the interactive login with Postman, but I guess it should be able to handle that.

**Reply by bmphjhu906** [reply]
Thank you for your reply.

**Comment by ilyashtrikul** [comment]
First request - authorize client
```
curl --request GET \
  --url 'https://zitadel.host/oauth/v2/authorize?response_type=code&client_id=your%40clientid&code_challenge=your_code_challenge&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fyour_redirect_uri&scope=openid%20profile%20email%20urn%3Azitadel%3Aiam%3Aorg%3Aid%3A215653578583329311&state=state'
```

You will receive a redirect response with location header like `/ui/login/login?authRequestID=215665000310655037`.
Following it you will be taken to the login form:
```
curl --request GET \
  --url 'https://zitadel.host[here Location header from response]'
```

In login form you should get login endpoint, CSRF token and authRequestID from login form, I uses xPath `//*/form/@action` for login endpoint, `//*/form/input[@name="gorilla.csrf.Token"]/@value` for csrf and  `//*/form/input[@name="authRequestID"]/@value` auth request.

Next send login form:
```
curl --request POST \
  --url https://zitadel.host[here login endpoint from login form] \
  --header 'Content-Type: multipart/form-data' \
  --form gorilla.csrf.Token=[here csrf token from login form] \
  --form authRequestID=[here auth request id from login form]\
  --form loginName=your_user_login
```

Next step - send password form, it is the same as login form, except you must also send the password
```
curl --request POST \
  --url https://zitadel.host[here login endpoint from password form]  \
  --header 'Content-Type: multipart/form-data' \
  --form gorilla.csrf.Token=[here csrf token from password form]\
  --form authRequestID=[here auth request id from password form] \
  --form loginName=your_user_login \
  --form password=your_user_password
```

If everything is OK, you will received another redirect to zitadel callback endpoint in location, follow it:
```
curl --request GET \
  --url '[url from location header in password form response]'
```

and and finally you will get a redirect to your `redirect_uri` with the authorization code, do not follow it, just get code param from request URI and send it to token endpoint:

```
curl --request POST \
  --url https://zitadel.host/oauth/v2/token \
  --header 'Content-Type: multipart/form-data' \
  --form code=[here code from uri] \
  --form grant_type=authorization_code \
  --form redirect_uri=[your redirect uri, it is the same as in first request] \
  --form code_verifier=[your code verifier, that you uses to generate code challenge] \
  --form client_id=your@clientid
```

That's all, for generating code_challenge and code_verifier you can use online tools like [this](https://tonyxu-io.github.io/pkce-generator/)

**Reply by bmphjhu906** [reply]
It helped me a lot. Thank you for your reply.👍
